lib/src/annotate.rs

```diff
@@ -0,0 +1,394 @@
+    }
+
+    if let Some(file_contents) = get_file_contents(store, file_path, tree)? {
+        file_cache.insert(commit_id.clone(), file_contents);
+    }
+
+    Ok(())
+}
+
+/// Get line by line annotations for a specific file path in the repo.
+/// If the file is not found, returns empty results.
+pub fn get_annotation_for_file(
+    repo: &dyn Repo,
+    starting_commit: &Commit,
+    file_path: &RepoPath,
+) -> Result<AnnotateResults, BackendError> {
+    if let Some(original_contents) =
+        get_file_contents(starting_commit.store(), file_path, &starting_commit.tree()?)?
+    {
+        let num_lines = original_contents.split_inclusive(|b| *b == b'\n').count();
+        let mut file_cache = HashMap::new();
+        file_cache.insert(starting_commit.id().clone(), original_contents.clone());
+
+        let original_line_map =
+            process_commits(repo, file_cache, starting_commit.id(), file_path, num_lines)?;
+
+        Ok(convert_to_results(original_line_map, &original_contents))
+    } else {
+        Ok(AnnotateResults {
+            file_annotations: Vec::new(),
+        })
+    }
+}
+
+/// Starting at the starting commit, compute changes at that commit relative to
+/// it's direct parents, updating the mappings as we go. We return the final
+/// original line map that represents where each line of the original came from.
+fn process_commits(
+    repo: &dyn Repo,
+    mut file_cache: FileCache,
+    starting_commit_id: &CommitId,
+    file_name: &RepoPath,
+    num_lines: usize,
+) -> Result<OriginalLineMap, BackendError> {
+    let predicate = RevsetFilterPredicate::File(FilesetExpression::file_path(file_name.to_owned()));
+    let revset = RevsetExpression::commit(starting_commit_id.clone())
+        .union(
+            &RevsetExpression::commit(starting_commit_id.clone())
+                .ancestors()
+                .filtered(predicate),
+        )
+        .evaluate_programmatic(repo)
+        .map_err(|e| match e {
+            RevsetEvaluationError::StoreError(backend_error) => backend_error,
+            RevsetEvaluationError::Other(_) => {
+                panic!("Unable to evaluate internal revset")
+            }
+        })?;
+    let mut commit_line_map = get_initial_commit_line_map(starting_commit_id, num_lines);
+    let mut original_line_map = HashMap::new();
+
+    for (cid, edge_list) in revset.iter_graph() {
+        let current_commit_line_map = commit_line_map.remove(&cid);
+        if let Some(current_commit_line_map) = current_commit_line_map {
+            let current_commit = repo.store().get_commit(&cid)?;
+            let (new_original_lines, new_commit_lines) = process_commit(
+                repo,
+                file_name,
+                &current_commit,
+                &mut file_cache,
+                current_commit_line_map,
+                &edge_list,
+            )?;
+            original_line_map.extend(new_original_lines);
+            commit_line_map.extend(new_commit_lines);
```

Comment:
better to pass `&mut commit_line_map` and `&mut origianl_line_map` in to `process_commit()` instead of allocating temporary `HashMap`.

Reply:
Oh, and maybe we shouldn't replace existing `commit_line_map` at fork point?

Reply:
I reverted the changes to original line map because that was excessive. I actually kind of like consuming the commit_line_map for the current commit map and returning the new parent's commit line maps so I left that in (so far). Let me know what you think.

Reply:
> I actually kind of like consuming the commit_line_map for the current commit map and returning the new parent's commit line maps 

I generally agree about that. However, in this case:
* the caller would have to merge existing `commit_line_map[id]` and `new_commit_lines[id]` line by line (at fork point, if I understand it correctly.)
* it's silly to hold temporary results in a `HashMap`